source/opt/fix_storage_class.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_storage_class.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status FixStorageClass::Process() {
+  bool modified = false;
+
+  get_module()->ForEachInst([this, &modified](Instruction* inst) {
+    if (FixInstruction(inst)) {
+      modified = true;
+    }
+  });
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool FixStorageClass::FixInstruction(Instruction* inst) {
+  switch (inst->opcode()) {
+    case SpvOpAccessChain:
+    case SpvOpPtrAccessChain:
+    case SpvOpInBoundsAccessChain:
+    case SpvOpCopyObject:
+    case SpvOpPhi:
+      return FixAddressInstruction(inst, 0);
+    case SpvOpSelect:
```

Comment:
I cannot understand why `SpvOpSelect` needs this.

Reply:
Even if DXC does not generate OpSelects now, it could be generated in the future or an optimization may generate it.  An `OpSelect %type %c %v1 %v2` is essentially the same as:

```
%v1 = ...
%v2 = ...
OpBranchConditional %c %l1 %l1
%l1 = OpLabel
OpBranch %merge
%l2 = OpLabel
OpBranch %merge
%merge = OpLabel
%v = OpPhi %type %v1 %l1 %v2 %l2
```

Notice it is the same as fixing up the `OpPhi` in the expanded coded.